src/Storages/ExternalTable/StorageExternalTable.cpp

```diff
@@ -0,0 +1,99 @@
+#include <Common/logger_useful.h>
+#include <Interpreters/Context.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Storages/ExternalTable/ClickHouse/ClickHouse.h>
+#include <Storages/ExternalTable/ExternalTableFactory.h>
+#include <Storages/ExternalTable/StorageExternalTable.h>
+
+namespace DB
+{
+
+StorageExternalTable::StorageExternalTable(
+        std::unique_ptr<ExternalTableSettings>  settings,
+        const StorageFactory::Arguments & args)
+: IStorage(args.table_id)
+, WithContext(args.getContext()->getGlobalContext())
+{
+    external_table = ExternalTableFactory::instance().getExternalTable(args.table_id.getTableName(), std::move(settings));
+
+    /// First, setStorageMetadata should be allowed to fail (the only failable part is getTableStructure function call), otherwise it will block Proton from starting up.
+    /// Second, when it fails, the exception should be caught, otherwise, Proton will fail to start.
+    /// TODO we could use cache to save the table structure, so that when Proton restarts it could read from the cache directly.
+    try
+    {
+        setStorageMetadata(args);
+    }
+    catch (const Exception & e)
+    {
+        LOG_ERROR(&Poco::Logger::get("ExternalTable-ClickHouse" + args.table_id.getFullTableName()),
+                  "Failed to fetch table structure, error: {}. Will keep retrying in background", e.what());
+        background_jobs.scheduleOrThrowOnError([this](){
```

Comment:
Since this is async, do we need wait for it done it construtor of StorageExternalTable?

Reply:
If we wait, seems we don't need make `background_jobs` as a data member.

Reply:
Sync to get table structure during create seems makes more sense. We have a bool flag `attach` to differentiate if this external table is an existing one load from file system or the initial creation DDL request.

Reply:
I actually forgot to check `attach`. So the logic should be this ( I commented in the code ):
```cpp
    /// Two situations:
    /// * Create a new table. In this case, we want it fails the create query if it fails to fetch the columns description. So that users know that there is something with the connection and they can try again once the issue is resolved.
    /// * Attach a table (Proton restarts). In this case, even it fails to fetch the columns description, we want to make sure that:
    ///   - it does not terminate Proton, otherwise Proton will never start again
    ///   - it does not block Proton from starting, otherwise Proton will get stuck
    ///   So, we let it keep retrying in the background, and hoping it will eventually succeeded (until the user drops the table).
```